observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingChannelInterceptor.java

```diff
@@ -16,33 +16,176 @@
 
 package io.grpc.observability.interceptors;
 
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Durations;
 import io.grpc.CallOptions;
 import io.grpc.Channel;
 import io.grpc.ClientCall;
 import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.Deadline;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
 import io.grpc.Internal;
+import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.interceptors.LogHelper.LogSinkWriter;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventLogger;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
 
-/** A logging interceptor for {@code LoggingChannelProvider}. */
+/**
+ * A logging interceptor for {@code LoggingChannelProvider}.
+ */
 @Internal
 public final class InternalLoggingChannelInterceptor implements ClientInterceptor {
+  private final LogSinkWriter writer;
 
   public interface Factory {
     ClientInterceptor create();
   }
 
   public static class FactoryImpl implements Factory {
+    Sink sink;
+
+    public FactoryImpl(Sink sink) {
+      this.sink = sink;
+    }
 
     @Override
     public ClientInterceptor create() {
-      return new InternalLoggingChannelInterceptor();
+      return new InternalLoggingChannelInterceptor(sink);
     }
   }
 
+  private InternalLoggingChannelInterceptor(Sink sink) {
+    this.writer = new LogSinkWriter(sink, TimeProvider.SYSTEM_TIME_PROVIDER);
```

Comment:
`LogSinkWriter` just uses a `Sink` and a `TimeProvider` and stores no other state and has no logic specific to this `InternalLoggingChannelInterceptor`. In that case it doesn't make any sense to create a `LogSinkWriter` per interceptor but just create one instance in the factory and have the `Factory.create` just supply it to all the interceptors.

Reply:
Unless of course `LogSinkWriter` will be enhanced (in future) to become stateful and/or will use more instance variables that will be dependent on each interceptor. What are those requirements?

Reply:
Updated to create one instance inside `InternalLoggingChannelInterceptor.FactoryImpl` and passing it to all the interceptors using `Factory.create`.